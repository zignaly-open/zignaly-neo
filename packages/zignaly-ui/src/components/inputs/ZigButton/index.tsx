import React from "react";
import { LoadingButton, LoadingButtonProps } from "@mui/lab";
import { ButtonGroup, styled, Tooltip } from "@mui/material";
import { dark } from "../../../theme";

export type ZigButtonProps = LoadingButtonProps & {
  ctaId?: string;
  tooltip?: string;
  active?: boolean;
};

const ZigButton = ({ active, tooltip, ctaId, color, ...props }: ZigButtonProps) => {
  const button = (
    <LoadingButton
      data-tack-cta={ctaId}
      {...props}
      // hack to preserve old behavior but allow for normal mui theming
      color={props.variant === "outlined" && !color ? "secondary" : color}
      className={active ? "MuiButton-active" : ""}
    />
  );
  return tooltip ? (
    <Tooltip title={tooltip}>
      {/* if we want to have a tooltip and have the button be disabled, we need a new container */}
      {props.disabled ? <span>{button}</span> : button}
    </Tooltip>
  ) : (
    button
  );
};

export const ZigButtonGroup: React.FC = styled(ButtonGroup)`
  .MuiButton-root {
    border-right-width: 0 !important;

    &:last-child {
      border-right-width: 1px !important;
    }

    &:hover,
    &.MuiButton-active {
      z-index: 3;
      box-shadow: 1px 0 0 ${dark.neutral400};
    }

    &.Mui-disabled {
      border-color: rgba(255, 255, 255, 0.1);
    }
  }
`;

export default ZigButton as typeof LoadingButton & typeof ZigButton;
