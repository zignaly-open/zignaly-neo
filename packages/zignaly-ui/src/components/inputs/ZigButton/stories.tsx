import React, { useState } from "react";
import { Meta, Story } from "@storybook/react";
import ZigButton, { ZigButtonProps } from "./index";
import { ButtonGroup, ButtonGroupProps } from "@mui/material";
import { styled } from "@mui/system";
import { Add } from "@mui/icons-material";
import ZigTypography from "../../display/ZigTypography";

export default {
  title: "Inputs/ZigButton",
  component: ZigButton,
  argTypes: {
    children: {
      type: "string",
    },
    disabled: {
      type: "boolean",
      defaultValue: false,
    },
    variant: {
      control: {
        type: "select",
      },
      options: ["outlined", "contained"],
    },
    tooltip: {
      type: "string",
    },
    size: {
      control: {
        type: "select",
      },
      options: ["small", "medium", "large"],
    },
  },
} as Meta<ZigButtonProps>;

const Template: Story<ZigButtonProps> = (args) => <ZigButton {...args} />;

export const Default = Template.bind({});
Default.args = {
  children: "Amount to Withdraw",
  tooltip: "Hello",
  variant: "outlined",
};

const TemplateGroup: Story<ZigButtonProps> = ({ size, ...args }) => {
  const [value, setValue] = useState<number>(1);
  return (
    <ButtonGroup variant="outlined" size={size as ButtonGroupProps["size"]}>
      <ZigButton active={value === 1} onClick={() => setValue(1)} variant="outlined" {...args} />
      <ZigButton active={value === 2} onClick={() => setValue(2)} variant="outlined" {...args} />
      <ZigButton active={value === 3} onClick={() => setValue(3)} variant="outlined" {...args} />
    </ButtonGroup>
  );
};

export const Group = TemplateGroup.bind({});
Group.args = {
  children: "Amount to Withdraw",
};

const Wrapper = styled("div")`
  & > * {
    display: inline-block;
    margin-right: 15px !important;
  }
`;

const TemplateTextButton: Story<typeof ZigButton> = (args) => {
  return (
    <>
      <ZigTypography variant={"h1"}>
        Hello <ZigButton variant={"text"} {...args} />
      </ZigTypography>
      <ZigTypography variant={"h2"}>
        Hello <ZigButton variant={"text"} {...args} />
      </ZigTypography>
      <ZigTypography variant={"h3"}>
        Hello <ZigButton variant={"text"} {...args} />
      </ZigTypography>
      <ZigTypography variant={"h4"}>
        Hello <ZigButton variant={"text"} {...args} />
      </ZigTypography>
      <ZigTypography variant={"body1"} component={"p"}>
        Hello, let me tell you a story of a button that lived in a very very long text. Hello, let
        me tell you a story of a button that lived in a very very long text. Hello, let me tell you
        a story of a button that lived in a very very long text. Yes, an{" "}
        <ZigButton variant={"text"} {...args} /> lived in a very long text and then died unhappily
        ever after. Yes, that was story of a button that lived in a very very long text.
      </ZigTypography>

      <ZigTypography variant={"body1"} component={"p"}>
        Hello, let me tell you a story of a button that lived in a very very long text. Hello, let
        me tell you a story of a button that lived in a very very long text. Yes, a{" "}
        <ZigButton startIcon={<Add />} endIcon={<Add />} variant={"text"}>
          Button
        </ZigButton>{" "}
        can have icons, imagine that.
      </ZigTypography>

      <ZigTypography variant={"body1"} component={"p"}>
        Hello, let me tell you a story of a button that lived in a very very long text. Hello, let
        me tell you a story of a button that lived in a very very long text. Yes, a{" "}
        <ZigButton href="https://xfuturum.com" variant={"text"}>
          Button
        </ZigButton>{" "}
        can be a link, imagine that.
      </ZigTypography>

      <ZigTypography variant={"body2"} component={"p"}>
        Hello <ZigButton variant={"text"} {...args} /> button in text in a very long
      </ZigTypography>
    </>
  );
};

const TemplateOutlinedButton: Story<typeof ZigButton> = (args) => {
  return (
    <Wrapper>
      <ZigButton variant={"outlined"} size="small" {...args} />
      <ZigButton variant={"outlined"} size="medium" {...args} />
      <ZigButton variant={"outlined"} size="large" {...args} />
      <ZigButton variant={"outlined"} size="xlarge" {...args} />
    </Wrapper>
  );
};

const TemplateLoadingButtons: Story<typeof ZigButton> = (args) => {
  return (
    <>
      <Wrapper>
        <ZigButton variant={"outlined"} loading size="small" {...args} />
        <ZigButton variant={"outlined"} loading size="medium" {...args} />
        <ZigButton variant={"outlined"} loading size="large" {...args} />
        <ZigButton variant={"outlined"} loading size="xlarge" {...args} />
      </Wrapper>
      <Wrapper>
        <ZigButton variant={"contained"} loading size="small" {...args} />
        <ZigButton variant={"contained"} loading size="medium" {...args} />
        <ZigButton variant={"contained"} loading size="large" {...args} />
        <ZigButton variant={"contained"} loading size="xlarge" {...args} />
      </Wrapper>
      <Wrapper>
        <ZigTypography variant={"body1"} component={"p"}>
          Please welcome the{" "}
          <ZigButton loading variant={"text"}>
            Button
          </ZigButton>
          !
        </ZigTypography>
      </Wrapper>
    </>
  );
};

const TemplateContainedButton: Story<typeof ZigButton> = (args) => {
  return (
    <Wrapper>
      <ZigButton variant={"contained"} size="small" {...args} />
      <ZigButton variant={"contained"} size="medium" {...args} />
      <ZigButton variant={"contained"} size="large" {...args} />
      <ZigButton variant={"contained"} size="xlarge" {...args} />
    </Wrapper>
  );
};

const TemplateIconButton: Story<typeof ZigButton> = (args) => {
  return (
    <Wrapper>
      <ZigButton startIcon={<Add />} variant={"contained"} size="small">
        {args.children}
      </ZigButton>
      <ZigButton startIcon={<Add />} variant={"contained"} size="medium">
        {args.children}
      </ZigButton>
      <ZigButton startIcon={<Add />} variant={"contained"} size="large">
        {args.children}
      </ZigButton>
      <ZigButton startIcon={<Add />} variant={"contained"} size="xlarge">
        {args.children}
      </ZigButton>
    </Wrapper>
  );
};

const buttonExampleArgs = {
  children: "Amount to Withdraw",
};

export const TextButton = TemplateTextButton.bind({});
export const ContainedButton = TemplateContainedButton.bind({});
export const OutlinedButton = TemplateOutlinedButton.bind({});
export const Icon = TemplateIconButton.bind({});
export const Loading = TemplateLoadingButtons.bind({});

TextButton.args = buttonExampleArgs;
OutlinedButton.args = buttonExampleArgs;
ContainedButton.args = buttonExampleArgs;
Icon.args = buttonExampleArgs;
Loading.args = buttonExampleArgs;
