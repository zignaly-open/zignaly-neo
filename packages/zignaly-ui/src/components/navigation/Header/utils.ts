// TODO: Move this for config or theme folder.
export const headerConfig = {
  height: "52px",
};
