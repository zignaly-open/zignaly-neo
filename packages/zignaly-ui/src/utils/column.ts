import styled from "styled-components";

export const Column = styled.div<{
  gap?: number;
  justifyContent?: "center" | "end" | "start" | "space-between";
  alignItems?: "center" | "end" | "start" | "space-between";
  textAlign?: "end" | "center" | "start";
  height?: string;
  flex?: number;
  width?: string;
}>`
  display: flex;
  flex-direction: column;
  flex: ${(props) => props.flex};
  gap: ${(props) => props.gap}px;
  justify-content: ${(props) => props.justifyContent};
  align-items: ${(props) => props.alignItems};
  text-align: ${(props) => `${props.textAlign}`};
  height: ${(props) => `${props.height}`};
  width: ${(props) => `${props.width}`};
`;
