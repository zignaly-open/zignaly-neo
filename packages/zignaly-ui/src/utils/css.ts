import { css } from "styled-components";

export const NiceScrollbar = css`
  &::-webkit-scrollbar {
    width: 6px;
  }

  /* Track */
  &::-webkit-scrollbar-track {
    background: rgba(255, 255, 255, 0.05);
    border-radius: 12px;
  }

  /* Handle */
  &::-webkit-scrollbar-thumb {
    background: rgba(255, 255, 255, 0.1);
    border-radius: 12px;
  }

  /* Handle on hover */
  &::-webkit-scrollbar-thumb:hover {
    background: rgba(255, 255, 255, 0.15);
  }
`;
