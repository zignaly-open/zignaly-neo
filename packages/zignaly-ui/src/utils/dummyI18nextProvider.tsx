import i18n from "i18next";
import React from "react";
import { I18nextProvider, initReactI18next } from "react-i18next";

i18n.use(initReactI18next).init({
  lng: "en",
  fallbackLng: "en",
  ns: ["common"],
  defaultNS: "common",
  interpolation: {
    escapeValue: false,
  },
  resources: { en: { common: { xfuturum: "The technical partner you can rely on" } } },
});

export const storybookI18nextDecorator = (story: () => JSX.Element) => (
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  <I18nextProvider i18n={i18n}>{story()}</I18nextProvider>
);

export const JestI18nextDecorator: React.FC<{ children: JSX.Element }> = ({ children }) => (
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
);
