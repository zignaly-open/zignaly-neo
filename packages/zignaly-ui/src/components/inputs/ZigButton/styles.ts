import { Box, styled } from "@mui/material";
import { NiceScrollbar } from "utils/css";
import React from "react";

type Props = { error?: string; width?: number };

export const StyledSelectWrapper = styled(Box)<Props>`
  .MuiInput-root {
    &.Mui-disabled {
      cursor: not-allowed;
      border-color: ${({ theme }) => theme.palette.neutral700};
    }

    &.Mui-focused {
      border-color: ${({ theme }) => theme.palette.neutral400};
    }

    &.Mui-error,
    &.Mui-error.Mui-focused {
      border-color: ${({ theme }) => theme.palette.redGraphOrError};

      .MuiInputLabel-root {
        color: ${({ theme }) => theme.palette.neutral200};
      }
    }
  }

  .MuiInput-input {
    background: transparent;
    border: none;
    color: rgb(193, 193, 200) !important;
    outline: none;
    font-weight: 400;
    font-size: 16px;
    line-height: 20px;
    letter-spacing: 0.55px;
    width: 100%;
    font-family: "Avenir Next", sans-serif;
    box-shadow: none !important;
    resize: none;
    background: linear-gradient(90deg, rgb(16 18 37) 0%, rgb(16 18 37) 35%, rgb(16 18 37) 100%);

    &:-webkit-autofill {
      -webkit-box-shadow: 0 0 0 1000px white inset;
      -webkit-background-clip: text;
    }

    &:-webkit-autofill:focus {
      -webkit-box-shadow: 0 0 0 50px white inset;
      -webkit-text-fill-color: #333;
    }

    -webkit-text-fill-color: #838b95 !important;

    &::placeholder {
      -webkit-text-fill-color: ${({ theme }) => theme.palette.neutral400} !important;
    }

    &.Mui-disabled {
      cursor: not-allowed;
      opacity: 0.67;
      color: ${({ theme }) => theme.palette.neutral100} !important;
    }
  }

  .zig-react-select {
    &__control {
      border: 1px solid
        ${({ theme, error }) => (error ? theme.palette.redGraphOrError : theme.palette.neutral600)};
      padding: 11px 24px 11px 16px;
      margin-top: 4px;
      min-height: 60px;
      border-radius: 5px;
      display: flex;
      align-items: center;
      background: rgba(16, 18, 37);
      background: linear-gradient(90deg, rgb(16 18 37) 0%, rgb(16 18 37) 35%, rgb(16 18 37) 100%);
      transition: border-color 0.2s;
      margin-bottom: 3px;

      ${({ width }) => width && `width: ${width}${width?.toString().includes("%") ? "" : "px"}`};

      padding-right: 0;

      &:hover {
        border-color: ${({ theme, error }) =>
          error ? theme.palette?.redGraphOrError : theme.palette.neutral400};
      }

      &--is-focused {
        border: 1px solid
          ${({ theme, error }) =>
            error ? theme.palette.redGraphOrError : theme.palette.neutral400};
        box-shadow: none !important;
      }
    }

    &__placeholder,
    &__single-value {
      padding: 0;
      font-size: 16px;
      line-height: 20px;
    }

    &__single-value,
    &__input {
      color: ${({ theme }) => theme.palette.neutral100} !important;
    }

    .zig-react-select__menu {
      border: 1px solid ${({ theme }) => theme.palette.neutral600} !important;
      color: ${({ theme }) => theme.palette.neutral200} !important;
      background: rgba(16, 18, 37) !important;
    }

    &__placeholder {
      color: ${({ theme }) => theme.palette.neutral400} !important;
    }

    &__menu {
      ${({ width }) => width && `width: ${width}px`};
      background: rgba(16, 18, 37);
      border: 1px solid ${({ theme }) => theme.palette.neutral600} !important;
      color: ${({ theme }) => theme.palette.neutral200} !important;

      &-list {
        ${NiceScrollbar.toString()};
      }
    }

    &__option--is-focused {
      cursor: pointer;
      background: rgba(255, 255, 255, 0.1);
    }

    &__option--is-selected {
      background: rgba(255, 255, 255, 0.2);
    }

    &__indicator {
      ${({ width }) => width && width <= 100 && `padding: 0 2px`};
    }
  }
` as React.FC<Props>;
