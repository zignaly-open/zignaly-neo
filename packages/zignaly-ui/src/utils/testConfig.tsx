import { render } from "@testing-library/react";
import { ThemeProvider as ThemeProviderMui } from "@mui/material";
import { ThemeProvider, ThemeProviderMui as ThemeInheritorMui } from "../index";
import React from "react";
import { getZignalyThemeExport } from "../theme";
import mediaQuery from "css-mediaquery";
import { JestI18nextDecorator } from "./dummyI18nextProvider";

export const renderWithProvidersUi = (ui: JSX.Element, renderOptions = {}) => {
  const { mui, legacyStyledComponentsDoNotUse } = getZignalyThemeExport();
  return render(ui, {
    wrapper: ({ children }: { children: JSX.Element }) => (
      <JestI18nextDecorator>
        <ThemeProvider theme={legacyStyledComponentsDoNotUse}>
          <ThemeInheritorMui theme={mui}>
            <ThemeProviderMui theme={mui}>{children}</ThemeProviderMui>
          </ThemeInheritorMui>
        </ThemeProvider>
      </JestI18nextDecorator>
    ),
    ...renderOptions,
  });
};

const createMatchMedia = (width: number) => {
  return (query: string) => {
    return {
      matches: mediaQuery.match(query, { width }),
      media: "",
      addListener: () => {},
      removeListener: () => {},
      onchange: () => {},
      addEventListener: () => {},
      removeEventListener: () => {},
      dispatchEvent: () => true,
    };
  };
};

export const resizeScreenSize = (width: number) => {
  window.matchMedia = createMatchMedia(width);
};
