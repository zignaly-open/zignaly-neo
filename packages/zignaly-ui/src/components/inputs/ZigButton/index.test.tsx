import React from "react";
import ZigButton from ".";
import { renderWithProvidersUi } from "../../../utils/testConfig";
import { fireEvent, waitFor } from "@testing-library/react";

describe("components/inputs/ZigButton", () => {
  describe("renders", () => {
    it("ZigButton small size", async () => {
      const { container } = renderWithProvidersUi(<ZigButton size={"small"} />);

      const button = container.querySelector("button") as Element;
      expect(button).toBeVisible();
      expect(button).toHaveClass("MuiButton-sizeSmall");
      const buttonStyles = getComputedStyle(button);
      expect(buttonStyles?.minWidth).toBe("76px");
      expect(buttonStyles?.minHeight).toBe("30px");
      expect(button).toMatchSnapshot();
    });
    it("ZigButton medium size", async () => {
      const { container } = renderWithProvidersUi(<ZigButton size={"medium"} />);

      const button = container.querySelector("button") as Element;
      expect(button).toBeVisible();
      expect(button).toHaveClass("MuiButton-sizeMedium");
      const buttonStyles = getComputedStyle(button);
      expect(buttonStyles?.minWidth).toBe("76px");
      expect(buttonStyles?.minHeight).toBe("36px");
      expect(button).toMatchSnapshot();
    });
    it("ZigButton large size", async () => {
      const { container } = renderWithProvidersUi(<ZigButton size={"large"} />);

      const button = container.querySelector("button") as Element;
      expect(button).toBeVisible();
      expect(button).toHaveClass("MuiButton-sizeLarge");
      const buttonStyles = getComputedStyle(button);
      expect(buttonStyles?.minWidth).toBe("110px");
      expect(buttonStyles?.minHeight).toBe("48px");
      expect(button).toMatchSnapshot();
    });
    it("ZigButton xlarge size", async () => {
      const { container } = renderWithProvidersUi(<ZigButton size={"xlarge"} />);

      const button = container.querySelector("button") as Element;
      expect(button).toBeVisible();
      expect(button).toHaveClass("MuiButton-sizeXlarge");
      const buttonStyles = getComputedStyle(button);
      expect(buttonStyles?.minWidth).toBe("127px");
      expect(buttonStyles?.minHeight).toBe("60px");
    });
    it("ZigButton contained variant by default", async () => {
      const { container } = renderWithProvidersUi(<ZigButton />);

      const button = container.querySelector("button") as Element;
      expect(button).toBeVisible();
      expect(button).toHaveClass("MuiButton-contained");
      const buttonStyles = getComputedStyle(button);
      expect(buttonStyles?.backgroundColor).not.toBe("transparent");
    });
    it("ZigButton outlined variant", async () => {
      const { container } = renderWithProvidersUi(<ZigButton variant={"outlined"} />);

      const button = container.querySelector("button") as Element;
      expect(button).toBeVisible();
      expect(button).not.toHaveClass("MuiButton-contained");
      expect(button).not.toHaveClass("MuiButton-contained");
      expect(button).toHaveClass("MuiButton-outlined");
      const buttonStyles = getComputedStyle(button);
      expect(buttonStyles?.backgroundColor).toBe("transparent");
    });
    it("ZigButton text variant", async () => {
      const { container } = renderWithProvidersUi(<ZigButton variant={"text"} />);

      const button = container.querySelector("button") as Element;
      expect(button).toBeVisible();
      expect(button).not.toHaveClass("MuiButton-contained");
      expect(button).not.toHaveClass("MuiButton-outlined");
      expect(button).toHaveClass("MuiButton-text");
      const buttonStyles = getComputedStyle(button);
      expect(buttonStyles?.backgroundColor).toBe("transparent");
    });
  });
  describe("disable", () => {
    it("ZigButton active test", async () => {
      const onClick = jest.fn();
      const { container } = renderWithProvidersUi(<ZigButton disabled={false} onClick={onClick} />);
      const button = container.querySelector("button") as Element;
      expect(button).not.toHaveClass("Mui-disabled");
      fireEvent.click(button);
      await waitFor(() => {
        expect(onClick).toHaveBeenCalledTimes(1);
      });
    });
    it("ZigButton disabled test", async () => {
      const onClick = jest.fn();
      const { container } = renderWithProvidersUi(<ZigButton disabled onClick={onClick} />);
      const button = container.querySelector("button") as Element;
      expect(button).toHaveClass("Mui-disabled");
      fireEvent.click(button);
      await waitFor(() => {
        expect(onClick).toHaveBeenCalledTimes(0);
      });
    });
  });
});
