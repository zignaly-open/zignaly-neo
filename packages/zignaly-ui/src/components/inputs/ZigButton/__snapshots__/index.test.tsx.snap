// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/inputs/ZigButton renders ZigButton large size 1`] = `
<button
  class="MuiLoadingButton-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonBase-root css-172u00w-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
  id=":r2:"
  tabindex="0"
  type="button"
/>
`;

exports[`components/inputs/ZigButton renders ZigButton medium size 1`] = `
<button
  class="MuiLoadingButton-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root css-z3a501-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
  id=":r1:"
  tabindex="0"
  type="button"
/>
`;

exports[`components/inputs/ZigButton renders ZigButton small size 1`] = `
<button
  class="MuiLoadingButton-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButtonBase-root css-25ezg8-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
  id=":r0:"
  tabindex="0"
  type="button"
/>
`;
